jb/itch5/compute_base.hpp

```diff
@@ -0,0 +1,247 @@
+ * Keep a collection of all the order books, and forward the right
+ * updates to them as it handles the different message types in
+ * ITCH-5.0.
+ *
+ * Uses:
+ * The order_book_depth object to get the book depth
+ * Virtual calls check_callback() to allow the implementation 
+ * of derivate callback signatures and call conditions
+ */
+
+class compute_base {
+ public:
+  //@{
+  /**
+   * @name Type traits
+   */
+  /// Define the clock used to measure processing delays.
+  typedef std::chrono::steady_clock::time_point time_point;
+  /// Callbacks
+  typedef std::function<void(time_point,
+			     message_header const&,
+			     stock_t const&)
+			> callback_type;
+  //@}
+
+  /**
+   * The collection of order book depths.
+   */
+  typedef std::unordered_map<stock_t,              /**<symbol, key      */
+			     order_book_depth,     /**<The order book for that symbol */
+			     boost::hash<stock_t>  /**<hash function    */
+			    > books_by_security;
+    
+  /// Initialize an empty handler
+  compute_base() = default;
+
+  /// Initialize an empty handler and its call condition
+  compute_base(callback_type const& _callback)
+    : callback_(_callback),
+      orders_(),
+      books_() {};
+
+  virtual ~compute_base() {};
+
+  /// Return the current timestamp for delay measurements
+  time_point now() const {
+    return std::chrono::steady_clock::now();
+  }
+
+ /**
+   * A convenient container for per-order data.
+   *
+   * Most market data feeds resend the security identifier and side
+   * with each order update, but ITCH-5.0 does not.  One needs to
+   * lookup that information based on the order id.  This literal type
+   * is used to keep that information around.
+   */
+  struct order_data {
+    stock_t stock;
+    buy_sell_indicator_t buy_sell_indicator;
+    price4_t px;
+    int qty;
+  };
+   
+ /** 
+   * Result of a order book modification.
+   * Used as return value on no_update message handlers
+   * to keep the state of the in progress almost_atomic
+   * order book modification
+   */
+  typedef std::tuple<bool,                         /**<true when is inside change, false otherwise       */
+		     order_data,                   /**<order data updated in the order book              */
+		     books_by_security::iterator   /**<iterator to the order book by security updated    */
+	            > update_result;
+
+  /**
+   * Verify callback condition.
+   */
+  void check_callback_condition(time_point ts,
+				message_header const& ms,
+				update_result const& ur,
+				bool ii = true);
+
+  /**
+   * Callback condition to be evaluated at derivated classes.
+   * It decides if the callback is actually invoked
+   * subject to its condition implementation.
+   */
+  virtual void call_callback_condition(time_point ts,
+				       message_header const& msg_header,
+				       stock_t const& stock,
+				       half_quote const& best_bid,
+				       half_quote const& best_offer,
+				       book_depth_t const book_depth,
```

Comment:
I do not think there would be any impact on performance.  Happy with whatever decision you make here.